### ISSUE TITLE
make tokio::run use Runtime::block_on_all

### ISSUE THREAD

**Issue by Keruspe** [tokio-rs/tokio#443]
Unless I'm missing something, it should be fairly trivial. it would allow running futures with types other than `()` and to get the corresponding `Result` as we used to get with `Core.run` in `tokio-core`.
Or maybe add a free `tokio::block_on_all` function instead?`

**Comment by jonhoo** [comment]
I think this should be pretty easily implementable using `Runtime::block_on` + `Runtime::shutdown_on_idle`. I suspect `tokio::run` doesn't do this because `Runtime::block_on` only recently landed (#391), but I'm in favor of changing it to allow non-`()` returns. If backwards compatibility is a major concern, I'd add `tokio::block_on_all` and maybe even go so far as to deprecate `tokio::run` in favor of it.

**Comment by carllerche** [comment]
I'm not sure how I feel about non-`()` returns right now. In theory it could work, but my original concern is that the behavior might not be obvious. Specifically, `run` will **not** unblock once the submitted future completes (it waits for all futures to complete). Thoughts?

**Comment by jonhoo** [comment]
I think that is mostly a naming concern. Specifically, we want to communicate that we are blocking on *all* futures, but returning the resolved item of the *given* future. Perhaps `resolve_and_wait`?

**Comment by carllerche** [comment]
No longer relevant. `tokio::run` is being removed in favor of a proc macro: `#[tokio::main]`.
